File: Sesion 11/Actividades/ActividadS11/ACTIVIDADS11.py
import csv

class Producto:
    def __init__(self, nombre="", precio=0.0, cantidad=0):
        if precio < 0 or cantidad < 0:
            raise ValueError("Precio y cantidad deben ser valores no negativos.")
        self.nombre = nombre
        self.precio = precio
        self.cantidad = cantidad

    def getNombre(self):
        return self.nombre

    def getPrecio(self):
        return self.precio

    def getCantidad(self):
        return self.cantidad

    def setNombre(self, nombre):
        self.nombre = nombre

    def setPrecio(self, precio):
        if precio < 0:
            raise ValueError("El precio no puede ser negativo.")
        self.precio = precio

    def setCantidad(self, cantidad):
        if cantidad < 0:
            raise ValueError("La cantidad no puede ser negativa.")
        self.cantidad = cantidad

    def mostrarInfo(self):
        print(f"Nombre: {self.nombre}, Precio: {self.precio}, Cantidad: {self.cantidad}")

    def leerDesdeCSV(self, linea):
        datos = linea.split(',')
        self.nombre = datos[0]
        self.precio = float(datos[1])
        self.cantidad = int(datos[2])

    def escribirEnCSV(self):
        return f"{self.nombre},{self.precio},{self.cantidad}\n"


def cargarIndice(indice, nombreArchivo):
    try:
        with open(nombreArchivo, 'r') as archivo:
            reader = csv.reader(archivo)
            pos = archivo.tell()
            for linea in reader:
                producto = Producto()
                producto.leerDesdeCSV(','.join(linea))
                indice[producto.getNombre()] = pos
                pos = archivo.tell()
    except FileNotFoundError:
        raise RuntimeError("No se pudo abrir el archivo.")


def agregarProducto(nombreArchivo, producto, indice):
    with open(nombreArchivo, 'a', newline='') as archivo:
        writer = csv.writer(archivo)
        pos = archivo.tell()
        writer.writerow([producto.getNombre(), producto.getPrecio(), producto.getCantidad()])
        indice[producto.getNombre()] = pos


def buscarProducto(nombreArchivo, indice, nombre):
    if nombre in indice:
        with open(nombreArchivo, 'r') as archivo:
            archivo.seek(indice[nombre])
            reader = csv.reader(archivo)
            linea = next(reader)
            producto = Producto()
            producto.leerDesdeCSV(','.join(linea))
            return producto
    else:
        raise RuntimeError("Producto no encontrado.")


def actualizarProducto(nombreArchivo, indice, nombre, nuevoPrecio, nuevaCantidad):
    if nombre in indice:
        with open(nombreArchivo, 'r+', newline='') as archivo:
            archivo.seek(indice[nombre])
            writer = csv.writer(archivo)
            writer.writerow([nombre, nuevoPrecio, nuevaCantidad])
            indice[nombre] = archivo.tell()
    else:
        raise RuntimeError("Producto no encontrado.")


def generarInforme(nombreArchivo):
    try:
        with open(nombreArchivo, 'r') as archivo:
            reader = csv.reader(archivo)
            print("Informe del inventario:")
            for linea in reader:
                producto = Producto()
                producto.leerDesdeCSV(','.join(linea))
                producto.mostrarInfo()
    except FileNotFoundError:
        raise RuntimeError("No se pudo abrir el archivo.")


def main():
    indice = {}
    nombreArchivo = "inventario.csv"

    try:
        # Crear archivo de inventario y cargar índice
        cargarIndice(indice, nombreArchivo)

        # Agregar productos al inventario
        try:
            agregarProducto(nombreArchivo, Producto("Producto1", 10.0, 100), indice)
            agregarProducto(nombreArchivo, Producto("Producto2", 20.0, 50), indice)
        except ValueError as e:
            print(f"Error al agregar producto: {e}")

        # Buscar y mostrar un producto
        try:
            productoEncontrado = buscarProducto(nombreArchivo, indice, "Producto1")
            productoEncontrado.mostrarInfo()
        except RuntimeError as e:
            print(f"Error al buscar producto: {e}")

        # Actualizar información de un producto
        try:
            actualizarProducto(nombreArchivo, indice, "Producto1", 12.0, 90)
        except RuntimeError as e:
            print(f"Error al actualizar producto: {e}")

        # Generar informe
        generarInforme(nombreArchivo)
    except RuntimeError as e:
        print(f"Error del sistema: {e}")


if __name__ == "__main__":
    main()

File: Sesion 11/Actividades/ActividadS11/ACTIVIDADS11.java
import java.io.*;
import java.util.HashMap;
import java.util.Map;

class Producto {
    private String nombre;
    private double precio;
    private int cantidad;

    public Producto() {
        this.nombre = "";
        this.precio = 0.0;
        this.cantidad = 0;
    }

    public Producto(String nombre, double precio, int cantidad) throws IllegalArgumentException {
        if (precio < 0 || cantidad < 0) {
            throw new IllegalArgumentException("Precio y cantidad deben ser valores no negativos.");
        }
        this.nombre = nombre;
        this.precio = precio;
        this.cantidad = cantidad;
    }

    // Getters
    public String getNombre() {
        return nombre;
    }

    public double getPrecio() {
        return precio;
    }

    public int getCantidad() {
        return cantidad;
    }

    // Setters
    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public void setPrecio(double precio) throws IllegalArgumentException {
        if (precio < 0) {
            throw new IllegalArgumentException("El precio no puede ser negativo.");
        }
        this.precio = precio;
    }

    public void setCantidad(int cantidad) throws IllegalArgumentException {
        if (cantidad < 0) {
            throw new IllegalArgumentException("La cantidad no puede ser negativa.");
        }
        this.cantidad = cantidad;
    }

    // Método para mostrar información del producto
    public void mostrarInfo() {
        System.out.println("Nombre: " + nombre + ", Precio: " + precio + ", Cantidad: " + cantidad);
    }

    // Método para leer un producto desde una línea CSV
    public void leerDesdeCSV(String linea) {
        String[] datos = linea.split(",");
        this.nombre = datos[0];
        this.precio = Double.parseDouble(datos[1]);
        this.cantidad = Integer.parseInt(datos[2]);
    }

    // Método para escribir un producto en formato CSV
    public String escribirEnCSV() {
        return nombre + "," + precio + "," + cantidad + "\n";
    }
}

public class Inventario {
    public static void cargarIndice(Map<String, Long> indice, String nombreArchivo) throws IOException {
        try (BufferedReader archivo = new BufferedReader(new FileReader(nombreArchivo))) {
            String linea;
            long pos = 0;
            while ((linea = archivo.readLine()) != null) {
                Producto producto = new Producto();
                producto.leerDesdeCSV(linea);
                indice.put(producto.getNombre(), pos);
                pos = archivo.getFilePointer();
            }
        } catch (FileNotFoundException e) {
            throw new IOException("No se pudo abrir el archivo.");
        }
    }

    public static void agregarProducto(String nombreArchivo, Producto producto, Map<String, Long> indice) throws IOException {
        try (RandomAccessFile archivo = new RandomAccessFile(nombreArchivo, "rw")) {
            archivo.seek(archivo.length());
            long pos = archivo.getFilePointer();
            archivo.writeBytes(producto.escribirEnCSV());
            indice.put(producto.getNombre(), pos);
        }
    }

    public static Producto buscarProducto(String nombreArchivo, Map<String, Long> indice, String nombre) throws IOException {
        if (indice.containsKey(nombre)) {
            try (RandomAccessFile archivo = new RandomAccessFile(nombreArchivo, "r")) {
                archivo.seek(indice.get(nombre));
                String linea = archivo.readLine();
                Producto producto = new Producto();
                producto.leerDesdeCSV(linea);
                return producto;
            }
        } else {
            throw new IOException("Producto no encontrado.");
        }
    }

    public static void actualizarProducto(String nombreArchivo, Map<String, Long> indice, String nombre, double nuevoPrecio, int nuevaCantidad) throws IOException {
        if (indice.containsKey(nombre)) {
            try (RandomAccessFile archivo = new RandomAccessFile(nombreArchivo, "rw")) {
                archivo.seek(indice.get(nombre));
                archivo.writeBytes(nombre + "," + nuevoPrecio + "," + nuevaCantidad + "\n");
                indice.put(nombre, archivo.getFilePointer());
            }
        } else {
            throw new IOException("Producto no encontrado.");
        }
    }

    public static void generarInforme(String nombreArchivo) throws IOException {
        try (BufferedReader archivo = new BufferedReader(new FileReader(nombreArchivo))) {
            String linea;
            System.out.println("Informe del inventario:");
            while ((linea = archivo.readLine()) != null) {
                Producto producto = new Producto();
                producto.leerDesdeCSV(linea);
                producto.mostrarInfo();
            }
        } catch (FileNotFoundException e) {
            throw new IOException("No se pudo abrir el archivo.");
        }
    }

    public static void main(String[] args) {
        Map<String, Long> indice = new HashMap<>();
        String nombreArchivo = "inventario.csv";

        try {
            // Crear archivo de inventario y cargar índice
            cargarIndice(indice, nombreArchivo);

            // Agregar productos al inventario
            try {
                agregarProducto(nombreArchivo, new Producto("Producto1", 10.0, 100), indice);
                agregarProducto(nombreArchivo, new Producto("Producto2", 20.0, 50), indice);
            } catch (IllegalArgumentException e) {
                System.err.println("Error al agregar producto: " + e.getMessage());
            }

            // Buscar y mostrar un producto
            try {
                Producto productoEncontrado = buscarProducto(nombreArchivo, indice, "Producto1");
                productoEncontrado.mostrarInfo();
            } catch (IOException e) {
                System.err.println("Error al buscar producto: " + e.getMessage());
            }

            // Actualizar información de un producto
            try {
                actualizarProducto(nombreArchivo, indice, "Producto1", 12.0, 90);
            } catch (IOException e) {
                System.err.println("Error al actualizar producto: " + e.getMessage());
            }

            // Generar informe
            generarInforme(nombreArchivo);
        } catch (IOException e) {
            System.err.println("Error del sistema: " + e.getMessage());
        }
    }
}
